### PR TITLE
[OpenCL] Implementation improvements

### PR DESCRIPTION
**Pull request by lukeiwanski**
OpenCL implementation improvements (#22)

- [Build] Use gcc/g++ as a host compiler to avoid #8394 (#54)
- [Eigen] Version bump
- [OpenCL] Implementation improvements
  - Register SYCL implementations for random ops
    - Simplify by using Eigen math functions
  - Registers Scatter and ScatterNd Ops for SYCL
  - Registers Stack op for SYCL
  - Fixes No sycl buffer found error for debug ops
  - Registers MatMul and Transpose Ops to SYCL device for double
  - Extends analyzer_cli_test.py test to cover SYCL
  - Fixes Transpose Op for double when on SYCL
  - Bumps Eigen version to fix double precision issue on SYCL
  - Extends SessionDebugTestBase to cover SYCL
  - Bumps Eigen Version
  - Refactors Ops registration
  - Introduces workaround for Const Op related to the difference between
   CUDA which uses pointers and OpenCL that uses buffers/accessors
  - Extends memory types to cover DEVICE_SYCL as well
  - Introduces  GetSYCLDevice() method that returns list of supported devices
   with GPU device having the highest priority ( doesn't include blacklisted devices )
  - ::internal::Transpose -> tensorflow::internal::Transpose in order to
   avoid compilation reported error
  - Adds sycl_runtime to bazels ARRAY_DEPS
  - Replicates TF_CALL_GPU_PROXY_TYPES for SYCL
  - Fixes an issue caused by switch to aligned allocator for sycl buffer (#53)
  - Fix testSimple and testConst in stack_op_test (#3)
  - RandomGamma has no GPU friendly implementation (#57)
  - Register batch normalization kernels for OpenCL (#61)
  - Compatibility fixes for TensorFlow 1.1.0-rc1
  - Fixes Scatter Op
  - Implements BatchMatmul Op for SYCL
  - Lowercase the device name when GPU or SYCL returned
  - kernel_estimator_test.py assertEqual-> assertAlmostEqual due to floating point 
    representation on the device

### CODE REVIEW

**Comment by lukeiwanski** [comment]
@ville-k, @Zakor94 can you confirm?

**Comment by drpngx** [comment]
Sometimes googlebot gets confused, could you `git pull --rebase` and push again?

**Comment by ville-k** [comment]
@lukeiwanski @googlebot  ok with me.

**Comment by Zakor94** [comment]
@lukeiwanski @googlebot yes ok with me too.

**Comment by yifeif** [comment]
@tensorflow-jenkins test this please

**Comment by yifeif** [comment]
@tensorflow-jenkins test this please

**Comment by yifeif** [comment]
@tensorflow-jenkins test this please

**Comment by yifeif** [comment]
@tensorflow-jenkins test this please

**Comment by drpngx** [comment]
Jenkins, test this please.

File: tensorflow/contrib/memory_stats/python/kernel_tests/memory_stats_ops_test.py [hunk 1]
```diff
@@ -32,7 +32,7 @@ class MemoryStatsOpsTest(test_util.TensorFlowTestCase):
   def testBytesLimit(self):
     # AllocatorStats.bytes_limit is set to zero for CPU allocators, so we skip
     # the check.
-    if not test.is_gpu_available():
+    if not test.is_gpu_available() or test_util.is_sycl_enabled():
```
Comment (benoitsteiner) [inline tensorflow/contrib/memory_stats/python/kernel_tests/memory_stats_ops_test.py]: Please delete all the test_util.is_sycl_enabled(). The point of having test is to expose problems, so hiding them is defeating the purpose.

File: tensorflow/core/common_runtime/direct_session_with_tracking_alloc_test.cc [hunk 2]
```diff
@@ -138,7 +138,8 @@ TEST(DirectSessionWithTrackingAllocTest, CostModelWarmup) {
   DirectSession* ds = static_cast<DirectSession*>(session.get());
   CostModelManager::CostModelMap cost_models;
   ds->ExportCostModels(&cost_models);
-  CHECK_EQ(cost_models.size(), 1);
+  ASSERT_GE(2, cost_models.size());
```
Comment (benoitsteiner) [inline tensorflow/core/common_runtime/direct_session_with_tracking_alloc_test.cc]: Why is this needed ?

**Review by benoitsteiner — CHANGES_REQUESTED** [review]

File: tensorflow/core/common_runtime/direct_session_with_tracking_alloc_test.cc (hunk 2 above)
Comment (lukeiwanski) [inline tensorflow/core/common_runtime/direct_session_with_tracking_alloc_test.cc]: Fixed in: https://github.com/tensorflow/tensorflow/pull/9117/commits/73e405572a59ba8ab1d92dc3f7cddff5c9fb240c 

That should not be there. Auto merge conflict resolution cannot be trusted.

File: tensorflow/contrib/memory_stats/python/kernel_tests/memory_stats_ops_test.py (hunk 1 above)
Comment (lukeiwanski) [inline tensorflow/contrib/memory_stats/python/kernel_tests/memory_stats_ops_test.py]: Removed in https://github.com/tensorflow/tensorflow/pull/9117/commits/73e405572a59ba8ab1d92dc3f7cddff5c9fb240c

The intention behind disabling the failing tests was to establish a regression baseline so that we can establish a smooth development process. We would enable more tests as we add the corresponding features, while being able to continually test that we haven't introduced regressions.

**Comment by drpngx** [comment]
Jenins, test this please.

**Comment by vrv** [comment]
@lukeiwanski these massive PRs are hard to review since there is a stream of new commits being added, and the PR is already very large.

We typically try to stage these changes in small PRs that are more easily reviewable -- can you break this change into smaller requests in some way?  As it is, I don't see anybody being able to safely review a change of this magnitude (117 files!)

**Comment by lukeiwanski** [comment]
@vrv Ok I will try. What is desired changeset size?

**Comment by vrv** [comment]
Benoit would have a better idea of what he'd like, but it's much easier to review a change that has like, <10 files, < 200-300 LoC change at a time; that we he can review it, it will have fewer things to change, and we can make more iterative progress.

Might be worth you and Benoit talking offline to see what's going to work well!

(I'm just on sync rotation this week and trying to help move things along :)

**Comment by lukeiwanski** [comment]
Appriciate and thanks @vrv !
My plan for this PR is to revert refactoring changes ( that should knock off 77 files ) and re-introduce them in a new one ( or series ) - I will check with @benoitsteiner on how to proceed.

**Comment by lukeiwanski** [comment]
@vrv Is current size of the pull requst acceptable or should I keep spliting it?

File: tensorflow/core/common_runtime/sycl/sycl_device_factory.cc [hunk 3]
```diff
@@ -18,24 +18,34 @@ limitations under the License.
...
+
+    size_t n = 1;
     auto iter = options.config.device_count().find("SYCL");
     if (iter != options.config.device_count().end()) {
       n = iter->second;
     }
+
     for (int i = 0; i < n; i++) {
-      string name = strings::StrCat(name_prefix, "/device:SYCL:", i);
+      string name = strings::StrCat(name_prefix, "/sycl:", i);
```
Comment (benoitsteiner) [inline tensorflow/core/common_runtime/sycl/sycl_device_factory.cc]: Can yo revert this change ? All devices should be named /device:DEVICE_NAME:DEVICE_ID (with the exception of CPUs and GPUs, which can also be referred to as /cpu:cpu_id and /gpu:gpu_id in order to preserve backward compatibility).

File: tensorflow/core/kernels/constant_op.cc [hunk 4]
```diff
@@ -184,6 +210,43 @@ class FillOp : public OpKernel {
...
+    OP_REQUIRES(
+        context, IsLegacyVector(Tdims.shape()),
+        errors::InvalidArgument("dims must be a vector of int32, got shape ",
+                                Tdims.shape().DebugString()));
+    const Tensor& Tvalue = context->input(1);
+
+    T data_host;
+    auto device = context->eigen_sycl_device();
+    auto size = sizeof(T);
+    auto src_ptr = GetBase(&Tvalue);
```
Comment (benoitsteiner) [inline tensorflow/core/kernels/constant_op.cc]: Handling of device pointers should really be done by the sycl compiler and in eigen, in order to avoid opening a can or worms in TensorFlow.

**Review by benoitsteiner — CHANGES_REQUESTED** [review]

File: tensorflow/core/common_runtime/sycl/sycl_device_factory.cc (hunk 3 above)
Comment (lukeiwanski) [inline tensorflow/core/common_runtime/sycl/sycl_device_factory.cc]: can DEVICE_NAME be lower case?

Comment (lukeiwanski) [inline tensorflow/core/common_runtime/sycl/sycl_device_factory.cc]: Fixed in https://github.com/lukeiwanski/tensorflow/commit/c4b1480529f03f17f16914d15e0e0576d480cf57

File: tensorflow/core/kernels/constant_op.cc (hunk 4 above)
Comment (lukeiwanski) [inline tensorflow/core/kernels/constant_op.cc]: We are aware of the issue and working on resolving it. I removed above workaround in https://github.com/lukeiwanski/tensorflow/commit/1a1525bae9a7e57a3faa2ed8e8a70de9dff3747f where I am registering Fill operation with .HostMemory("value"). Is that acceptable for time being?

**Comment by rmlarsen** [comment]
@benoitsteiner could you please take another look?

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by lukeiwanski** [comment]
The MacOs CPU Tests seems to be failing due to a missing dependency in gif_archive..

ERROR: /private/var/tmp/_bazel_jenkins/13e370a18c169b19baeafefb05212b85/external/gif_archive/BUILD.bazel:8:1: undeclared inclusion(s) in rule '@gif_archive//:gif':
this rule is missing dependency declarations for the following files included by 'external/gif_archive/lib/openbsd-reallocarray.c':
 
Is that caused by our changeset or is it present in the upstream? I cannot see how we could cause it.

**Comment by rmlarsen** [comment]
@lukeiwanski no, the build error on MacOS is unrelated to your changes.

File: tensorflow/core/kernels/constant_op.cc (hunk 4 above)
Comment (lukeiwanski) [inline tensorflow/core/kernels/constant_op.cc]: 8ba7dbdbcf9dc1e540a47988437093e2d2fd2420 seems to be more appropriate solution

File: tensorflow/contrib/kernel_methods/python/kernel_estimators_test.py [hunk 5]
```diff
@@ -92,7 +92,7 @@ def testLinearlySeparableBinaryDataNoKernels(self):
     # Since the data is linearly separable, the classifier should have small
     # loss and perfect accuracy.
     self.assertLess(metrics['loss'], 0.1)
-    self.assertEqual(metrics['accuracy'], 1.0)
+    self.assertAlmostEqual(metrics['accuracy'], 1.0, 7)
```
Comment (gunan) [inline tensorflow/contrib/kernel_methods/python/kernel_estimators_test.py]: According to the comment above, I am worried about almost equal here. I know 7 digits is very very precise, but still we are making the test looser.

**Comment by gunan** [comment]
Jenkins, test this please.

File: configure [hunk 6]
```diff
@@ -683,7 +683,7 @@ if [ "$TF_NEED_OPENCL" == "1" ]; then
 while true; do
   fromuser=""
   if [ -z "$HOST_CXX_COMPILER" ]; then
-    default_cxx_host_compiler=$(which clang++-3.6 || true)
+    default_cxx_host_compiler=$(which g++-4.8 || true)
```
Comment (lukeiwanski) [inline configure]: We locked to GCC-4.8 as it was default installation on Ubuntu 14.04.
This has been changed in 50197b9792e3eb262201ee143cc82546a151e388 and d777320c6a9b730449f744a5ad8d068da12953ed
However, on ubuntu 16.04 user will have to add "-D_GLIBCXX_USE_CXX11_ABI=0" manually to the command line if there is custom compiler installed on the machine.

File: tensorflow/contrib/kernel_methods/python/kernel_estimators_test.py (hunk 5 above)
Comment (lukeiwanski) [inline tensorflow/contrib/kernel_methods/python/kernel_estimators_test.py]: The reason for that was the fact that sometimes OpenCL represent this value as 0.99999996
Reverting in 1a6045241a27228ba817cb0b722679c357077e36

**Review by benoitsteiner — APPROVED** [review]

**Comment by benoitsteiner** [comment]
Jenkins, test this please.
